core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -44,11 +44,40 @@ import org.apache.kafka.common.utils.Time
 import org.apache.kafka.common.{IsolationLevel, TopicPartition, Uuid}
 import org.apache.kafka.metadata.LeaderRecoveryState
 import org.apache.kafka.server.common.MetadataVersion
-import org.apache.kafka.storage.internals.log.{AppendOrigin, FetchDataInfo, FetchIsolation, FetchParams, LogOffsetMetadata}
+import org.apache.kafka.storage.internals.log.{AppendOrigin, FetchDataInfo, FetchIsolation, FetchParams, LogOffsetMetadata, LogOffsetsListener}
 
 import scala.collection.{Map, Seq}
 import scala.jdk.CollectionConverters._
 
+/**
+ * Listener receives notification from an Online Partition.
+ *
+ * A listener can be (re-)registered to an Online partition only. The listener
+ * is notified as long as the partition remains Online. When the partition fails
+ * or is deleted, respectively `onFailed` or `onDeleted` are called once. No further
+ * notifications are sent after this point on.
+ *
+ * Note that the callbacks are executed in the thread that triggers the change
+ * AND that locks may be held during their execution. They are meant to be used
+ * as notification mechanism only.
+ */
+trait PartitionListener {
+  /**
+   * Called when the Log increments its high watermark.
+   */
+  def onHighWatermarkUpdated(partition: TopicPartition, offset: Long): Unit = {}
+
+  /**
+   * Called when the Partition has a failure (e.g. goes offline).
+   */
+  def onFailed(partition: TopicPartition): Unit = {}
+
+  /**
+   * Called when the Partition is deleted.
+   */
+  def onDeleted(partition: TopicPartition): Unit = {}
```

Comment:
I wonder if we can clarify in the name what the deletion signifies. I think it just means the replica was deleted, so maybe `onReplicaDeleted` or something like that perhaps?

Reply:
I kind of like `onDeleted` because it goes well with `onFailed`. Isn't it implicit that we are talking about the local partition (== replica)? I could perhaps make this clearer in the javadoc.

Reply:
Is there another word we could use like "removed"? Maybe that's also unclear but I remember in the discussion about bumping epoch on stop replica there are a lot of reasons a replica is no longer on a broker. 😅

Reply:
This was is basically called when `Partition.delete` is called so `delete` makes sense in my opinion. But, yeah, naming is hard :).